Annotated documents proposal

alpmestan opened the pull request:
@goldfirere @bgamari This is part 2 of the plan. Feedback welcome!

alpmestan committed:
first draft of the annotated documents proposal

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,300 @@
+    newtype SDoc a = SDoc { runSDoc :: SDocContext -> Doc a }
+
+and would come with similar instances as ``Doc`` (since with this definition,
+``SDoc ~ ReaderT SDocContext Doc``), except for ``Foldable`` and
```
I think you mean `Coercible` here, not `~`.

alpmestan replied:
Yes, it was a "handwavy `~`", which indeed corresponds to `Coercible`.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,300 @@
+the documents, even supporting embedding ASTs from all 3 GHC passes.::
+
+    data PassWrapper f
+      = WrapperPs (f GhcPs) -- parser AST
```
This representation suggests that one error message might have both e.g. `HsExpr GhcPs` and `HsExpr GhcRn` in the same message. I doubt we want that. So I'm unconvinced here.

alpmestan replied:
Yes, this was just a quick idea to easily be able to embed AST fragments from all 3 phases.

On the other hand, do we really want to parametrize `ErrAnnotation` as a whole by the pass that the annotations belong to, yielding documents of type `SDoc (ErrAnnotation GhcTc)` and so on? 
We would have to make the pass "disappear" from the type at some point to be able to uniformly process documents with annotations from the different phases. One way is to wrap `ErrAnnotation`, another is to wrap the entire document. The former is almost what I suggested (except that ErrAnnotation would be the "wrappee" this time), the latter seems more intrusive since we would have to take that "wrapped document" type in many places instead of `SDoc`s.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,300 @@
+    (>>=) :: SDoc a -> (a -> SDoc b) -> SDoc b
+
+The ``Monad`` is exactly what is required to substitute all annotations by some
+annotation-free document that depends on the annotation value. In our scenario,
```
What's "our scenario" here?

alpmestan replied:
The scenario where we "substitute all annotations by some annotation-free documents that depends on the annotation value". But the wording isn't great, I'll find a better way to bring home the point about `>>=` being exactly many annotated document consumers will want to use, to produce the original error document or a fancier version of it that makes use of the annotations values in an interesting way (yielding an annotation-free document in both cases).

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,300 @@
+      | Column (Int -> Doc a) -- new!
+      | ...
+
+While this now prevents ``Doc`` from being ``Foldable`` or ``Traversable``
```
Wait -- I'm confused. The text above said that we could support `Foldable` and `Traversable`. Here it says we can't. Which is it? I do think that `Foldable` would be useful: it would allow an IDE to e.g. extract all the types eagerly from an error message and then do some processing -- without rendering the message. I don't think `Foldable` is a must-have, if it's hard to support, but I do think it would be useful.

Actually, without a `Foldable` instance (or similar), how would an IDE extract the information it wants? I guess it would have to pattern-match on the `Doc` type directly.

Or, we would have a function

```hs
extractAnnotations :: SDocContext -> SDoc a -> [AnnotationInContext a]
extractAnnotations = ...

data AnnotationInContext a
  = AIC { annot :: a, annotPrintedRow :: Int, annotPrintedColBegin :: Int, annotPrintedColEnd :: Int }
```

or some such. This way, the tool gets all the annotations with the information it needs to syntax-highlight etc. the right bits of the output. Of course, the tool could also ignore the row/col information if appropriate.

Maybe this was all clear to you, but I think this could be spelled out more in the proposal.

alpmestan replied:
Your suggestion is quite interesting in its own right, but for now I'll just clarify my reasoning/idea and then we can decide which way we want the proposal to go.

The basic idea is that when we extend our existing `Doc` with that new `| Pure a` constructor, what we have _is_ `Foldable` and `Traversable`. At that point, we can use `toList` to collect annotations and `Traversable` to applicatively transform them. But when we add our extra constructors for introducing sub-documents that depend on the current column number, nesting level and what not, we lose those instances. The statement about having `Foldable` and `Traversable` only applies to the version of `Doc` that the proposal has at that point of the document, not the final version with the extra few constructors. I can look into making this clearer.

Note that `Foldable` and `Traversable` were already out of the window for `SDoc`, right from the start. So the idea I had was simply to offer a function like `collect :: SDoc a -> [a]`, where `collect` assumes that the document doesn't make use of the `Column`/`Nesting`/... constructors, and clearly states so.

As far as I can tell, our inspiration for this approach -- [wl-pprint-extras](http://hackage.haskell.org/package/wl-pprint-extras-3.5.0.5/docs/Text-PrettyPrint-Free.html#t:SimpleDoc) -- gets away by defining a `SimpleDoc` type, that doesn't have those `Column`/etc constructors nor some more typical document constructors that `Doc` has, and maps `Doc`s to `SimpleDoc`s, which are then displayed. Interestingly, `SimpleDoc` still supports annotations, but the scoped kind: `| Effect e (SimpleDoc e)`.

goldfirere replied:
Ah -- this helps. But actually, I think we will want to be able to output the annotations with location information, as IDEs will often want that. So maybe both functions are warranted: one that can get the annotations directly, and one that requires e.g. ribbon width to get them but comes with output location info.

Incidentally, if we have `collect :: SDoc a -> [a]`, then surely we can write a `Foldable` instance. Maybe it's best not to, if we're worried about invariants being broken. But perhaps we expose an interface that does not allow end users (that is, all of GHC outside the `Outputable` module) to make location-dependent annotations. Then, the `Foldable` instance would be safe to use.

alpmestan replied:
Well, what would we do in the clause for the `| Column (Int -> Doc a)` constructor of `Doc`?

``` hs
collect :: Doc a -> [a]
...
collect (Column k) = collect (k 0)
...
```

?

That would not be right if `k` is, say, `\i -> if i /= 0 then Pure i else Empty`.

goldfirere replied:
If the public (that is, outside the `Outputable` module) interface to the pretty-printer does not allow the construction of an annotation hidden behind a condition dependent on a column number, then we can safely do just what you wrote above. That is, given a `Doc` `Column f`, it is an invariant that the result of `collect (f n)` does not depend on `n`.

alpmestan replied:
Your "if" is something that we will have to check while implementing this proposal, then, I suppose. It is hard to take this conversation further without more details about where/when those constructors need to be introduced or eliminated. In case the assumption doesn't hold, there isn't much we can do without a rendered document; any annotation that's only available for a given column number argument or something along those lines is a symptom of a bad document. Those column number/etc arguments should only be relevant for positioning the elements relatively to one another, not for determining their contents. So yes, actually, maybe we can just "cheat" here, and pass dummy column/etc numbers when we want to collect the annotations.

goldfirere replied:
Perhaps I've meant my "if" as: let's try to have it be so. Because we get nice properties as a result!

alpmestan replied:
Yes, it's not unreasonable to think we can succeed. I'll revise the document accordingly, to outline our reasoning around this property.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,300 @@
+
+    type MsgDoc = SDoc
+
+    data ErrDoc = ErrDoc {
```
Especially in light of the previous big-ADT-for-all-errors proposal, I'm not sure about the advantage of this three-part structure.

alpmestan replied:
The three-part structure is what we already have, and I thought it would be best to frame this proposal as completely separate from the error-values one, so that the two would effectively just "compose" if/when accepted and implemented. So I'm just talking about decorating the existing error infra with annotations, there. Should I just make the error-values proposal a requirement for this one instead? And presenting what would be the final result, of implementing the two proposals?

goldfirere replied:
OK -- happy to defer to you on this one.

goldfirere reviewed (commented):
Looks good. Perhaps this also might point to the significant previous discussion on the topic, e.g. from the GHC ticket #8809.

alpmestan commented:
Yes, I should definitely mention #8089 somewhere, I forgot, thanks!

alpmestan committed:
address feedback from Richard

alpmestan commented:
Updated the document to reflect your feedback.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,304 @@
+operations anyway and it is the type that we manipulate the most),
+it nicely solves the problem of having to "guess" properties about the
+textual rendition of an annotation, while mixing several pieces of documents
+to produce a larger one using specific semantics.
```
I still want more information about e.g. `collect`-like operations. How will clients access the annotations? Is it worth thinking/talking about returning annotations with source locations?

alpmestan replied:
I'm not done with that section. :-)

I'm waiting for the corresponding "sub-discussion" to come to a concrete conclusion that I can then expand into a proper explanation as well as examples.

goldfirere reviewed (commented):
This may be getting ready for public consumption. But it's very hard to assess this without seeing a larger example. For example, it would be really instructive to see how a real, from-the-wild GHC function could be rewritten with this new facility... along with a hypothetical consumer.

It might be good to present this as a chance for early feedback, instead of strictly a new proposal. (This is actually out of scope of the proposal process.) After the actual implementation starts, we could continue to review and refine.

alpmestan committed:
another pass over the whole proposal

alpmestan commented:
@goldfirere I just pushed a new commit which hopefully addresses all your feedback. Let me know what you think about this update, when you get a chance. :-)

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+    data SDocContext = SDC { ... }
+    newtype SDoc = SDoc { runSDoc :: SDocContext -> Doc }
+
+The exact contents of ``SDocContext`` is not very important, one just needs
```
Maybe just show `DynFlags` here?

alpmestan replied:
Well, the example at the end of the _Effects and interactions_ section uses the real definition. Did you get the impression that mentioning `SDocContext` adds a lot of "weight"?

goldfirere replied:
What I meant was to show that `SDocContext` includes `DynFlags`, among other things. I just thought it would help to illustrate what `SDocContext` is about.

alpmestan replied:
Perhaps I can just spell it out too, but not necessarily show the definition of `PprStyle` and what not. It's just one line to add.

goldfirere replied:
Sounds like a plan.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+
+This ``Monad`` instance lets us substitute all the annotations in a document
+by fresh, new subdocuments that can depend on the annotation values. Of
+particular interest is the case where ``b`` is ``()`` or ``Void``, indicating
```
But `()` and `Void` are different. I think we want `Void` here.

alpmestan replied:
Yes, `Void` is more accurate, let's only mention that one.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+
+Pieces of code that don't produce annotations in documents can either
+produce documents that are completely polymorphic in the annotation type,
+or ``SDoc Void`` values. The polymorphic variant has the merit of being
```
Consumers should probably work with `SDoc Void`; producers should probably produce `SDoc a`.

alpmestan replied:
Yes, that should be stated clearly, thanks!

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+If we can ensure that the public API for ``Doc`` doesn't allow users to build
+documents that may or may not have some annotations, depending e.g on
+the column number, we can implement ``toList :: Doc a -> [a]``
+or ``foldMap :: Monoid m => (a -> m) -> Doc a -> m``. This
```
Why can't we just write a `Foldable` instance then? This is surprising, especially coming so soon after deciding not to have `Foldable`.

alpmestan replied:
Right, I meant to imply that we could therefore have our `Foldable` instance, but I should make it crystal clear.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+``Couldn't match expected type`` error message is currently emitted by the
+following code, from ``compiler/typecheck/TcBinds.hs``.::
+
+  typeSigErrMsg :: SDoc
```
The code you're looking for is in TcErrors. This little chunk of code is used only in `COMPLETE` pragmas.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+The ``Outputable`` class in GHC lets us specify how to render values of all
+sorts of types as documents:::
+
+    class Outputable a where
```
I still favor

```
class Outputable ty annot where
  ppr :: ty -> SDoc annot
```

Then, we can have

```
instance Outputable Type TcErrorAnnot where
  ppr = embed
instance {-# OVERLAPPABLE #-} Outputable Type a where
  ppr = pprType
```

Otherwise, we'll have to change a **lot** of calls to `ppr` throughout the codebase to be annotation-aware.

alpmestan replied:
Hmm, right, this does seem to be the nicest option as this will not incur additional work for people who emit those annotations. I guess we could even have an `Outputable Type Void` instance (and possibly others) in place of the generic, overlappable one -- in case someone objects to using overlappable instances for this.

goldfirere replied:
As far as I understand it, overlapping instances are bad only when combined with orphan instances. I think this one is pretty innocent. And it's useful: I might want `ppr :: Type -> SDoc ParseError` somehow.

Another approach: Conveniently, we have `fmap Data.Void.absurd :: forall a. SDoc Void -> SDoc a`.  We could say `fromVoidDoc = fmap absurd` and then if someone wants to print a `Type` in a `SDoc ParseError`, they could use the `Outputable Type Void` instance and `fromVoidDoc`. That's not so bad. I *think* it's valid to have `{-# RULES "fromVoidDoc" fromVoidDoc = unsafeCoerce #-}` to avoid rebuilding the whole `SDoc` tree just for a type change.

Maybe we present both the design with the overlapping instance and without and see what people prefer. Regardless, `fromVoidDoc` seems like a grand idea.

alpmestan replied:
Let's indeed mention the options we discussed here and see where the discussion goes.

goldfirere commented on 0000-richer-documents.rst:
```diff
@@ -0,0 +1,460 @@
+The only aspect of the implementation that is not crystal clear at this point
+is the handling of annotation nodes in a few key functions from
+``compiler/utils/Pretty.hs``. Fortunately, any implementation that does not
+preserve the current layout bit for bit will quite likely be caught by the
```
Is this actually true? I thought the testsuite normally ignores whitespace changes.

alpmestan replied:
I'm thinking the "drifts" would be significant enough to cause more than that, perhaps shifting some contents to the next line. Worst case, I could tweak the driver, to make tests sensible to whitespace changes, temporarily. But you have a point, I should make my statement a bit more precise/honest.

goldfirere replied:
It indeed would be nice to be able to ask the testsuite for whitespace-sensitivity. Not just for this work, but in general, in case some other patch proves to be whitespace-sensitive in the future.

goldfirere reviewed (commented):
Looks good. Just a few smallish suggestions. Thanks!

alpmestan committed:
address more feedback

alpmestan commented:
Thanks for the new round of feedback, I hopefully addressed all of it in the latest update.

alpmestan commented:
@goldfirere Is this good to go for you?

goldfirere commented:
Yes, go for it. I have not done another review, but it's better to start getting public feedback and to keep this process rolling forward. Thanks!

alpmestan commented:
https://github.com/ghc-proposals/ghc-proposals/pull/307
